### DISCUSSION TITLE
Compilation Issue on Ubuntu

### DISCUSSION THREAD

**Discussion by BlueDream145** [trustwallet/wallet-core#3501] [Q&A] [answered]
**Environment:**

Operating System: Ubuntu 22.04.3 LTS
Trust Wallet Core Version: Just updated today
Rust Version: rustc 1.72.0-nightly (36fb58e43 2023-06-26)
Cargo Version: cargo 1.72.0-nightly (03bc66b55 2023-06-23)
Cbindgen Version: 0.20.0

**Issue Description:**

I'm encountering an issue when attempting to compile Trust Wallet Core on Ubuntu. The compilation process involves running a script called rust-bindgen to generate C bindings from Rust code. However, during this process, I encounter syntax errors when calling cbindgen.
`cbindgen --crate $CRATE --output ../src/rust/bindgen/$HEADER_NAME`

For example, I'm seeing errors like this:

```
ERROR: Parsing crate `wallet-core-rs`:`/home/user/wallet-core/rust/wallet_core_rs/src/ffi/bitcoin/legacy.rs`:
Error("expected `;`")
Try running `rustc -Z parse-only /home/user/wallet-core/rust/wallet_core_rs/src/ffi/bitcoin/legacy.rs` to see a nicer error message
ERROR: Couldn't generate bindings for /home/user/wallet-core/rust.
```

And when i run  `rustc -Z parse-only /home/user/wallet-core/rust/wallet_core_rs/src/ffi/bitcoin/legacy.rs`, i got absolutely no output so no error i guess.

I used to be able to manually fix these syntax issues in files like tw_keypair and tw_bitcoin, but now with the last updates the code is more complex, and I'm having difficulty adapting it for compilation.

**Steps to Reproduce:**

Navigate to the rust directory.
Run the rust-bindgen script.

**The script i used to install all dependencies on a fresh new ubuntu install**
```

#!/bin/bash

export DEBIAN_FRONTEND=noninteractive

sudo apt-get update
sudo apt-get install --no-install-recommends -y \
    ca-certificates curl file \
    build-essential \
    autoconf automake autotools-dev libtool xutils-dev build-essential

export SSL_VERSION=1.0.2u

sudo curl https://www.openssl.org/source/openssl-$SSL_VERSION.tar.gz -O
tar -xzf openssl-$SSL_VERSION.tar.gz
sudo rm openssl-$SSL_VERSION.tar.gz

cd openssl-$SSL_VERSION
sudo ./config
sudo make depend
sudo make install
cd ..
sudo rm -rf openssl-$SSL_VERSION*

export OPENSSL_LIB_DIR=/usr/local/ssl/lib
export OPENSSL_INCLUDE_DIR=/usr/local/ssl/include
export OPENSSL_STATIC=1

sudo apt-get install -y pkg-config ninja-build ccache \
    clang clang-14 llvm-14 libc++-dev libc++abi-dev apt-utils \
    ruby-full libboost-all-dev git cmake unzip snap wget

sudo apt-get install -y ed

sudo snap install rustup --classic

curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
export PATH=/root/.cargo/bin:$PATH

if [ -d "wallet-core" ]; then
    sudo rm -rf "wallet-core"
fi
git clone https://github.com/trustwallet/wallet-core
cd wallet-core
sudo ./tools/install-sys-dependencies-linux
sudo ./tools/install-dependencies
sudo ./tools/install-rust-dependencies
sudo sed -i '7 a #include "string.h"' src/Nebulas/Address.cpp
sudo update-alternatives --set c++ /usr/bin/clang++
sudo cmake -D CMAKE_CXX_COMPILER="/usr/bin/clang++-14"
sudo cargo install --force cbindgen
sudo ./tools/build-and-test
```

**Expected Behavior:**

The compilation process should complete successfully without syntax errors in the generated C headers.

**Actual Behavior:**

Syntax errors are encountered while generating C headers, preventing successful compilation.

**Additional Information:**

I have attempted to address these syntax issues manually, but I would like to understand why I'm unable to compile with the original syntax.

**Comment by BlueDream145** [comment]
I asked to GPT some help and he successfully fixed me some of the files, for example for `/home/user/wallet-core/rust/wallet_core_rs/src/ffi/bitcoin/legacy.rs`:


```
let Some(mut encoded) = CByteArrayRef::new(input, input_len).as_slice() else {
    return CUInt64Result::error(1);
};
```
This seems to be trying to use some form of error handling or pattern matching, but the syntax isn't valid Rust. The correct syntax for using the Option type in a conditional is with if let or match.

Here's a fixed version of the tw_bitcoin_legacy_calculate_transaction_fee function:
```

#[deprecated]
#[no_mangle]
pub unsafe extern "C" fn tw_bitcoin_legacy_calculate_transaction_fee(
    input: *const u8,
    input_len: usize,
    sat_vb: u64,
) -> CUInt64Result {
    if let Some(mut encoded) = CByteArrayRef::new(input, input_len).as_slice() {
        if let Ok(tx) = Transaction::consensus_decode(&mut encoded) {
            let weight = tx.weight();
            let fee = weight.to_vbytes_ceil() * sat_vb;
            CUInt64Result::ok(fee)
        } else {
            CUInt64Result::error(1)
        }
    } else {
        CUInt64Result::error(1)
    }
}
```

But after that i have more confusing errors, and i can't fix each file for each update...

**Comment by satoshiotomakan** [comment, marked as answer]
Hi @alexandre10044, thank you for opening the issue.
The `let-else` syntax you mentioned is valid, it was released in rust 1.65.

Could you please confirm if you can compile Rust part by typing
```bash
cd rust
cargo build
```

Please make sure if you do not override the Rust version via `rustup override` and no toolchain configs present neither in Cargo home nor in `./rust`.
The last I would suggest to update cbindgen to the latest version.

**Reply by apomonzer4-ship-it** [reply]
cd rust
cargo build

**Comment by BlueDream145** [comment]
@satoshiotomakan 

Thank you for your prompt response and providing the necessary details.

As per your request, I have tested the compilation of the Rust part by running cargo build, and I can confirm that it works as expected. There were no issues with the build.

Additionally, I have removed oldcbindgen: (0.20.0)

` apt install cbindgen`
(Apt installing me an old version, i don't know why)

And i have installed cbindgen to the latest version as suggested. Here are the commands I used:

`sudo cargo install --force cbindgen`
cbindgen v0.26.0` (executable `cbindgen`)

But now cbindgen can't be called from my terminal...

So now i'm getting this:

`cbindgen --crate wallet-core-rs --output ../src/rust/bindgen/WalletCoreRSBindgen.h
bash: /usr/bin/cbindgen: No such file or directory`

And for the override and toolchain configs, not sure if i'm doing something wrong, but this should give infos about the situation:

`rustup show
Default host: x86_64-unknown-linux-gnu
rustup home:  /home/user/.rustup

installed toolchains
--------------------

stable-2023-06-01-x86_64-unknown-linux-gnu
stable-x86_64-unknown-linux-gnu
nightly-2022-07-24-x86_64-unknown-linux-gnu
nightly-2023-06-27-x86_64-unknown-linux-gnu (default)
nightly-x86_64-unknown-linux-gnu

installed targets for active toolchain
--------------------------------------

wasm32-unknown-emscripten
wasm32-unknown-unknown
x86_64-unknown-linux-gnu

active toolchain
----------------

nightly-2023-06-27-x86_64-unknown-linux-gnu (default)
rustc 1.72.0-nightly (36fb58e43 2023-06-26)
`

**Comment by BlueDream145** [comment]
So now with cbindgen 0.26 successfully installed but rust-bindgen can't find cbindgen...

`sudo ./tools/rust-bindgen 
Generating Native target
    Finished release [optimized] target(s) in 3.12s
Generating WASM target
    Finished release [optimized] target(s) in 2.84s
./tools/rust-bindgen: line 113: cbindgen: command not found
`

Here what happen when i try to cbindgen manually:

`cbindgen --crate wallet-core-rs --output ../src/rust/bindgen/WalletCoreRSBindgen.h`

`WARN: Skip tw_coin_registry::BITCOIN - (not `pub`).
WARN: Skip tw_coin_registry::ETHEREUM - (not `pub`).
WARN: Skip tw_coin_registry::RONIN - (not `pub`).
WARN: Skip tw_bitcoin::PROTOCOL_ID - (not `pub`).
WARN: Skip tw_bitcoin::OPERATION - (not `pub`).
WARN: Skip tw_bitcoin::NO_LEAF_HASH - (not `pub`).
WARN: Skip tw_bitcoin::NO_CONTROL_BLOCK - (not `pub`).
WARN: Skip tw_bitcoin::NO_TAPROOT_PAYLOAD - (not `pub`).
WARN: Skip tw_keypair::SIGNATURE_LENGTH - (not `pub`).
WARN: Skip tw_keypair::R_LENGTH - (not `pub`).
WARN: Skip tw_keypair::S_LENGTH - (not `pub`).
WARN: Skip tw_keypair::SIGN_RETRIES - (not `pub`).
WARN: Skip tw_hash::OUTPUT_HASH_LEN_RANGE - (not `pub`).
WARN: Skip tw_hash::PERSONAL_INPUT_MAX_LEN - (not `pub`).
WARN: Missing `[defines]` entry for `feature = "bitcoin-legacy"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "bitcoin-legacy"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "bitcoin-legacy"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "bitcoin-legacy"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "bitcoin-legacy"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "bitcoin-legacy"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "bitcoin-legacy"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "bitcoin-legacy"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "bitcoin-legacy"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "bitcoin-legacy"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "bitcoin-legacy"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "bitcoin-legacy"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "bitcoin-legacy"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "bitcoin-legacy"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "ethereum-abi"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "ethereum-abi"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "ethereum-abi"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "ethereum-abi"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "ethereum-abi"` in cbindgen config.
WARN: Missing `[defines]` entry for `feature = "ethereum-rlp"` in cbindgen config.
`

**Comment by BlueDream145** [comment]
Ok this was an environment issue, thanks for yur help.

So for next people that have troubles like me,

Just ensure cbindgen is 0.26 and maybe edit rust-bindgen script to find the executable location,

For the last output i gave, they are just warnings, so we can skip them.

Thanks @satoshiotomakan
